### ISSUE TITLE
Internationalization

### ISSUE THREAD

**Issue by rzach** [Carnap/Carnap#271]
Probably not terribly high priority but would eventually be useful: 

- Put all the strings that Carnap displays into an xml file.
- Crowdsource translation of those strings into different languages
- Provide UI to select language Carnap should use

**Comment by lf-** [comment]
Yesod has some built in i18n functionality: https://www.yesodweb.com/book/internationalization

**Comment by gleachkr** [comment]
I think the relevant typeclass is provided by https://hackage.haskell.org/package/shakespeare-2.0.25/docs/Text-Shakespeare-I18N.html, and we already use shakespeare in the client, so we could use probably use the same translation file format on both sides.

**Comment by gleachkr** [comment]
645e3ec2eadd83e51b5939ea978845489486f17c adds initial internationalization support on the server side, using the Yesod approach. We can now begin moving text from into `messages\en.msg`, and invite translators to supply matching files for their own languages. I'll try to chip away at it, but pull requests are also very welcome (should be easy to understand what to do, after looking at the yesod book link that @lf- included).

**Comment by gleachkr** [comment]
Just to keep track of progress on this, here's a provisional list of what needs to be internationalized and what's done. Many of these files don't actually need any work, but they should at least be checked to make sure that any incidental user-facing text is properly internationalized.

- [ ] Foundation.hs
- [ ] Application.hs

## Templates

- [x] auth-layout.hamlet
- [x] carnap-book.hamlet 
- [x] default-layout.hamlet
- [x] default-layout-wrapper.hamlet
- [x] document.hamlet 
- [x] documentIndex.hamlet
- [X] homepage.hamlet
- [X] infopage.hamlet
- [ ] instructor.hamlet
- [x] passwordEntry.hamlet 
- [x] register.hamlet
- [x] review.hamlet
- [x] user.hamlet
- [ ] admin.julius
- [ ] assignment-state.julius
- [ ] command.julius
- [ ] createTrees.julius
- [ ] infopage.julius
- [ ] instructor.julius
- [ ] review.julius
- [ ] status-warning.julius
- [ ] user.julius

## Handlers

- [ ] Admin.hs
- [ ] API
- [ ] Assignment.hs
- [ ] Book.hs
- [ ] Chapter.hs
- [ ] Command.hs
- [ ] Common.hs
- [ ] Document.hs
- [ ] Home.hs
- [ ] Info.hs
- [ ] Instructor.hs
- [ ] Register.hs
- [ ] Review.hs
- [ ] Rule.hs
- [ ] Serve.hs
- [x] User.hs

## Utilities

- [ ] API.hs
- [ ] Assignment.hs
- [ ] Database.hs
- [ ] Data.hs
- [ ] Grades.hs
- [ ] Handler.hs
- [ ] LTI.hs

## Filters

- [ ] Anchors.hs
- [ ] CounterModelers.hs
- [ ] ProofCheckers.hs
- [ ] Qualitative.hs
- [ ] Randomize.hs
- [ ] RenderFormulas.hs
- [ ] Sequent.hs
- [ ] Sidenotes.hs
- [ ] SynCheckers.hs
- [ ] Translate.hs
- [ ] TreeDeduction.hs
- [ ] TruthTables.hs
- [ ] TruthTrees.hs
- [ ] Util.hs

**Comment by ismailserin** [comment]
While translating "messages\en.msg" into Turkish, I ran into a problem. Since the syntax of Turkish is different from English, the translations of the sentences that are separated into different lines cause problems. For example, HomepageText1-9. In order to get meaningful Turkish UI, it would be better to put the complete sentence in each line.

**Comment by diegofernandess** [comment]
While working on the file templates/user.hamlet I noticed there are a couple of tables generated by Handler/user.hs whose items are left in English (e.g. 'Assignment', 'Due Date', 'Description', etc.). Could anyone please indicate how to internationalize them also?

**Comment by gleachkr** [comment]
@diegofernandess - Thanks for helping out with this!

I think you'll find the stuff you're looking for around here:

https://github.com/Carnap/Carnap/blob/54a586444a42ed7802aa45b97691ae9535e9fc72/Carnap-Server/Handler/User.hs#L216

There are a few bits of embedded hamlet through the user.hs file that need to be internationalized. The syntax is the same as in the `.hamlet` files, but you find it in between quasi-quotes that look like `[whamlet|...|]`, or `[hamlet|...]`.

**Comment by diegofernandess** [comment]
Yeah, I had already tried to apply the instructions given above to this file. For the first whamlet block, returning assignment data, it worked, but for the second block, returning student data, I got a compile error.

For example, replacing

https://github.com/Carnap/Carnap/blob/54a586444a42ed7802aa45b97691ae9535e9fc72/Carnap-Server/Handler/User.hs#L297
with 

`<dt.col-sm-3> _{MsgBlabla}`

and adding the respective line in messages/en.msg, I get the compilation error:

```
[49 of 51] Compiling Handler.User     ( Handler/User.hs, dist/build/Handler/User.o, dist/build/Handler/User.dyn_o )

Handler/User.hs:295:16: error:
    • No instance for (RenderMessage site AppMessage)
        arising from a use of ‘getMessageRender’
    • In the second argument of ‘fmap’, namely ‘getMessageRender’
      In the first argument of ‘(>>=)’, namely
        ‘(fmap (toHtml GHC.Base..)) getMessageRender’
      In a stmt of a 'do' block:
        ((fmap (toHtml GHC.Base..)) getMessageRender
           >>=
             (\ urender_a9LTD
                -> (asWidgetT GHC.Base.. toWidget) (urender_a9LTD MsgBlabla)))
    |
295 |     = [whamlet|
    |                ^...
[50 of 51] Compiling Handler.API.Instructor.Students ( Handler/API/Instructor/Students.hs, dist/build/Handler/API/Instructor/Students.o, dist/build/Handler/API/Instructor/Students.dyn_o )
builder for '/nix/store/gdpaxcm4r9zl6mn3ywwrqpsdxgdwn883-Carnap-Server-0.0.0.drv' failed with exit code 1
error: build of '/nix/store/gdpaxcm4r9zl6mn3ywwrqpsdxgdwn883-Carnap-Server-0.0.0.drv' failed
make: *** [Makefile:30: build-production] Error 100
```

**Comment by gleachkr** [comment]
Ah, oops. I'm traveling right now, so I can't double check this easily, but could you try changing the type signature for `personalInfo`, here: https://github.com/Carnap/Carnap/blob/54a586444a42ed7802aa45b97691ae9535e9fc72/Carnap-Server/Handler/User.hs#L291
to  `personalInfo :: UserData -> Maybe Course -> WidgetFor App ()`?

**Comment by diegofernandess** [comment]
Yes, it works!

**Comment by diegofernandess** [comment]
What about popup messages, such as 
https://github.com/Carnap/Carnap/blob/54a586444a42ed7802aa45b97691ae9535e9fc72/Carnap-GHCJS/src/Carnap/GHCJS/Action/Translate.hs#L117
How could they be internationalized?

**Comment by gleachkr** [comment]
@diegofernandess Diego, thank you so much for the new internationalization contributions. This is extremely helpful. I'll get it merged ASAP.

About extending internationalization to popups in the GHCJS part - here's what I think needs to happen.

1. We need to create a `RenderMessage` instance that can be used on the client side. This can be done using `mkMessages` from `Text.Shakespeare.I18N` - [here's](https://www.yesodweb.com/book/internationalization) an explanation from the yesod book, and [here's](https://hackage.haskell.org/package/shakespeare-2.0.30/docs/Text-Shakespeare-I18N.html) the documentation for the relevant bit of Shakespeare.

2. We need to have a way of getting the language codes that the browser is requesting. So we need to wrap this interface using the GHCJS FFI: https://developer.mozilla.org/en-US/docs/Web/API/Navigator/language

3. We can then create a bunch of internationalization files (similar to the ones that already exist on the server side), use those to provide the translations, and insert them by calling `renderMessage` directly.

**Comment by gleachkr** [comment]
Merged and deployed!
